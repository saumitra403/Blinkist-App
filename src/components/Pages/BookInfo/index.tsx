import { Box } from "@material-ui/core";
import { useParams } from "react-router-dom";
import Typography from "../../atoms/Typography";
import Icon from "../../atoms/Icons";
import AccessTimeIcon from "@mui/icons-material/AccessTimeOutlined";
import { Button } from "@mui/material";
import ArrowForwardIcon from "@mui/icons-material/ArrowForward";
import Tabs from "../../molecules/Tabs";
import { useState } from "react";
import Img from "../../atoms/Images";
const BookInfo = ({ books,setBooks }: any) => {
  const tabData = [
    {
      value: 0,
      label: "Synopsis",
    },
    {
      value: 1,
      label: "Who's it for?",
    },
    {
      value: 2,
      label: "About the Author",
    },
  ];
  const [value, setValue] = useState(0);
  const handleClick = (val: any) => {
    setValue(0);
  };
  let { bookId } = useParams();
  let currData = books[bookId!];
  const handleClickFinished = async (id: any) => {
    //put request
    let index = bookId;
    currData.status = "Finished";
    currData.progress = 100;
    const book = books.filter((curr: any) => {
      return curr.id!=index;
    })
    book.splice(index,0,currData);
    setBooks(book)

    await fetch(`http://localhost:3000/books/${index}`, {
      method: "PUT",
      body: JSON.stringify(currData),
      headers: {
        "Content-Type": "application/json",
      },
    });
  };
  return (
    <Box
      sx={{
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
        height: "895px",
        width: "100%",
        marginTop: "86px",
      }}
    >
      <Box sx={{ height: "inherit", width: "inherit" }}>
        <Typography
          variant="body2"
          sx={{ position: "absolute", top: "146px", left: "264px" }}
        >
          Get The Key Ideas From
        </Typography>
        <Typography
          variant="h1"
          sx={{ position: "absolute", top: "206px", left: "264px" }}
        >
          {currData.bookname}
        </Typography>
        <Typography
          variant="subtitle2"
          sx={{ position: "absolute", top: "275px", left: "264px" }}
        >
          Turning Your Business into an Enduring Great Company
        </Typography>
        <Typography
          sx={{ position: "absolute", top: "324px", left: "264px" }}
        >{`by ${currData.author}`}</Typography>
        <Typography
          variant="caption"
          sx={{ position: "absolute", top: "363px", left: "296px" }}
        >
          15 minute-read
        </Typography>
        <Icon
          Icons={
            <AccessTimeIcon
              sx={{ position: "absolute", top: "362px", left: "266px" }}
            ></AccessTimeIcon>
          }
        ></Icon>
        <Box sx={{ display: "flex", justifyContent: "space-evenly", position: "absolute", top: "464px", left: "264px" }}>
          <Button
            variant="outlined"
            sx={{
              color: "#22C870",
              fontSize: "16px",
              border: "1px solid #042330 !important",
              width: "fit-content",
              height: "44px",
            }}
          >
            Read Now
          </Button>
          {
            currData.status==="Reading" ?
            <Button
              variant="contained"
              sx={{
                background: "#2CE080",
                height: "44px",
                color: "#03314B",
                marginLeft: "21px",
              }}
              onClick={handleClickFinished}
            >
              Finished Reading
            </Button> : <Box sx={{display: "none"}}></Box>
          }
          <Button
            variant="text"
            endIcon={<ArrowForwardIcon></ArrowForwardIcon>}
            sx={{
              color: "#6D787E",
              fontSize: "16px",
              height: "20px",
              width: "fit-content",
              marginTop: "12px",
              marginLeft: "31.5px",
            }}
          >
            Send To Kindle
          </Button>
        </Box>
        <Box sx={{ position: "absolute", top: "568px", left: "264px" }}>
          <Tabs tabData={tabData} stateHandler={handleClick}></Tabs>
          <Box>
            {value == 0 ? (
              <Typography variant="body2" sx={{width: "600px"}}>{`Beyond Entrepreneurship 2.0 (2020) updates Jim Collins and Bill Laziers essential 1992 business handbook, Beyond Entrepreneurship for the entrepreneurs, visionaries, and innovators of today. This new edition combines the timeless business advice and strategy of the original text, supplemented with cutting-edge insights and case studies pertinent to todays business world.`}</Typography>
            ) : (
              <Box sx={{display: "none"}}></Box>
            )}
          </Box>
        </Box>
        <Box sx={{ position: "absolute", top: "206px", left: "898px" }}>
          <Img src={currData.src} width="304px" height="304px"></Img>
        </Box>
      </Box>
    </Box>
  );
};

export default BookInfo;
