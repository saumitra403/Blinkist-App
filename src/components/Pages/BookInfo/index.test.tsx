import { render, screen } from "@testing-library/react";
import BookInfo from ".";
import { BrowserRouter } from "react-router-dom";
import { fireEvent } from "@testing-library/react";
let response, books: any;
const router = (books: any, setBooks: any, id: any) => {
  return (
    <BrowserRouter>
      <BookInfo books={books} setBooks={setBooks} id={id}></BookInfo>
    </BrowserRouter>
  );
};

const setBooks = jest.fn();
describe("Home", () => {
  let response, books: any;
  beforeAll(async () => {
    response = await fetch("http://localhost:3000/books");
    books = await response.json();
  });
  test("should render input", async () => {
    const tree = render(router(books, setBooks,1));
    await new Promise((r) => setTimeout(r, 2000));
    expect(tree).toMatchSnapshot();
  });
  test("should render selected tab 0", async () => {
    render(router(books, setBooks,1));
    await new Promise((r) => setTimeout(r, 2000));
    const tabsElement = screen.getByTestId(/tab-0/);
    fireEvent.click(tabsElement);
    expect(tabsElement).toHaveStyle("color:#2CE080");
  });
  test("should render tab 1", async () => {
    render(router(books, setBooks,1));
    await new Promise((r) => setTimeout(r, 2000));
    const tabsElement = screen.getByTestId(/tab-1/);
    fireEvent.click(tabsElement);
    // const progressBarElement = await screen.getByText(
    //   "Beyond Entrepreneurship 2.0 (2020) updates Jim Collins and Bill Laziers essential 1992 business handbook, Beyond Entrepreneurship for the entrepreneurs, visionaries, and innovators of today. This new edition combines the timeless business advice and strategy of the original text, supplemented with cutting-edge insights and case studies pertinent to todays business world."
    // );
    expect(tabsElement).toHaveStyle("color:#2CE080");
  });

  test("should render tab 2", async () => {
    render(router(books, setBooks,1));
    await new Promise((r) => setTimeout(r, 2000));
    const tabsElement = screen.getByTestId(/tab-2/);
    fireEvent.click(tabsElement);
    // const progressBarElement = await screen.getByText(
    //   "Beyond Entrepreneurship 2.0 (2020) updates Jim Collins and Bill Laziers essential 1992 business handbook, Beyond Entrepreneurship for the entrepreneurs, visionaries, and innovators of today. This new edition combines the timeless business advice and strategy of the original text, supplemented with cutting-edge insights and case studies pertinent to todays business world."
    // );
    expect(tabsElement).toHaveStyle("color:#2CE080");
  });

  test("Finished Reading Button Should not Be Displayed", async () => {
    render(router(books, setBooks,2));
    await new Promise((r) => setTimeout(r, 2000));
    const buttonElement = screen.queryByText("Finished Reading");
    expect(buttonElement).not.toBeInTheDocument();
  })

  test("Finished Reading Button Should not Be Displayed", async () => {
    render(router(books, setBooks,7));
    await new Promise((r) => setTimeout(r, 2000));
    const buttonElement = screen.queryByText("Finished Reading");
    expect(buttonElement).not.toBeInTheDocument();
  })
  
  test("Button is Clickable and state must be changed", async () => {
    render(router(books, setBooks,1));
    await new Promise((r) => setTimeout(r, 2000));
    const buttonElement = screen.getByText("Finished Reading");
    fireEvent.click(buttonElement);
    expect(buttonElement).not.toBeInTheDocument();
  })

});


//1(Reading),2(Finished),7(Null)