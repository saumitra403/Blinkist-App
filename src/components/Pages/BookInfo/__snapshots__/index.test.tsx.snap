// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render input 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-1ofxksp-MuiTypography-root"
            data-testid="typo"
          >
            Get The Key Ideas From
          </p>
          <h1
            class="MuiTypography-root MuiTypography-h1 css-gmx1yk-MuiTypography-root"
            data-testid="typo"
          >
            Beyond Entrepreneurship
          </h1>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-1ax2xp1-MuiTypography-root"
            data-testid="typo"
          >
            Turning Your Business into an Enduring Great Company
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1m2q8tz-MuiTypography-root"
            data-testid="typo"
          >
            by Jim Collins & Bill Lazier
          </p>
          <span
            class="MuiTypography-root MuiTypography-caption css-1indtro-MuiTypography-root"
            data-testid="typo"
          >
            15 minute-read
          </span>
          <div
            data-testid="box"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-f9sn5-MuiSvgIcon-root"
              data-testid="AccessTimeOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm.5-13H11v6l5.25 3.15.75-1.23-4.5-2.67z"
              />
            </svg>
          </div>
          <div
            class="MuiBox-root MuiBox-root-3"
          >
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1d7rih8-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Read Now
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-5te66f-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Finished Reading
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1x29clv-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Send To Kindle
              <span
                class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowForwardIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiBox-root MuiBox-root-4"
          >
            <div
              class="makeStyles-tabs-5 MuiBox-root css-1gsv261"
            >
              <div
                class="MuiTabs-root css-1dfm0b0-MuiTabs-root"
                data-testid="tabs"
              >
                <div
                  class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
                  style="overflow: hidden; margin-bottom: 0px;"
                >
                  <div
                    class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                    role="tablist"
                  >
                    <button
                      aria-selected="true"
                      class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary Mui-selected css-1xlyk3o-MuiButtonBase-root-MuiTab-root"
                      data-testid="tab-0"
                      role="tab"
                      tabindex="0"
                      type="button"
                    >
                      Synopsis
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                    <button
                      aria-selected="false"
                      class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-1xlyk3o-MuiButtonBase-root-MuiTab-root"
                      data-testid="tab-1"
                      role="tab"
                      tabindex="-1"
                      type="button"
                    >
                      Who's it for?
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                    <button
                      aria-selected="false"
                      class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-1xlyk3o-MuiButtonBase-root-MuiTab-root"
                      data-testid="tab-2"
                      role="tab"
                      tabindex="-1"
                      type="button"
                    >
                      About the Author
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <span
                    class="MuiTabs-indicator css-nuqw77-MuiTabs-indicator"
                    style="left: 0px; width: 0px;"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-6"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-1e74oyv-MuiTypography-root"
                data-testid="typo"
              >
                Beyond Entrepreneurship 2.0 (2020) updates Jim Collins and Bill Laziers essential 1992 business handbook, Beyond Entrepreneurship for the entrepreneurs, visionaries, and innovators of today. This new edition combines the timeless business advice and strategy of the original text, supplemented with cutting-edge insights and case studies pertinent to todays business world.
              </p>
            </div>
          </div>
          <div
            class="MuiBox-root MuiBox-root-7"
          >
            <img
              alt="not found"
              class=""
              data-testid="img"
              height="304px"
              src="/Assets/2.png"
              width="304px"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1ofxksp-MuiTypography-root"
          data-testid="typo"
        >
          Get The Key Ideas From
        </p>
        <h1
          class="MuiTypography-root MuiTypography-h1 css-gmx1yk-MuiTypography-root"
          data-testid="typo"
        >
          Beyond Entrepreneurship
        </h1>
        <h6
          class="MuiTypography-root MuiTypography-subtitle2 css-1ax2xp1-MuiTypography-root"
          data-testid="typo"
        >
          Turning Your Business into an Enduring Great Company
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1m2q8tz-MuiTypography-root"
          data-testid="typo"
        >
          by Jim Collins & Bill Lazier
        </p>
        <span
          class="MuiTypography-root MuiTypography-caption css-1indtro-MuiTypography-root"
          data-testid="typo"
        >
          15 minute-read
        </span>
        <div
          data-testid="box"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-f9sn5-MuiSvgIcon-root"
            data-testid="AccessTimeOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm.5-13H11v6l5.25 3.15.75-1.23-4.5-2.67z"
            />
          </svg>
        </div>
        <div
          class="MuiBox-root MuiBox-root-3"
        >
          <button
            class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1d7rih8-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Read Now
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-5te66f-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Finished Reading
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1x29clv-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Send To Kindle
            <span
              class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowForwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root MuiBox-root-4"
        >
          <div
            class="makeStyles-tabs-5 MuiBox-root css-1gsv261"
          >
            <div
              class="MuiTabs-root css-1dfm0b0-MuiTabs-root"
              data-testid="tabs"
            >
              <div
                class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
                style="overflow: hidden; margin-bottom: 0px;"
              >
                <div
                  class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                  role="tablist"
                >
                  <button
                    aria-selected="true"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary Mui-selected css-1xlyk3o-MuiButtonBase-root-MuiTab-root"
                    data-testid="tab-0"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    Synopsis
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-1xlyk3o-MuiButtonBase-root-MuiTab-root"
                    data-testid="tab-1"
                    role="tab"
                    tabindex="-1"
                    type="button"
                  >
                    Who's it for?
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-1xlyk3o-MuiButtonBase-root-MuiTab-root"
                    data-testid="tab-2"
                    role="tab"
                    tabindex="-1"
                    type="button"
                  >
                    About the Author
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <span
                  class="MuiTabs-indicator css-nuqw77-MuiTabs-indicator"
                  style="left: 0px; width: 0px;"
                />
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root MuiBox-root-6"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1e74oyv-MuiTypography-root"
              data-testid="typo"
            >
              Beyond Entrepreneurship 2.0 (2020) updates Jim Collins and Bill Laziers essential 1992 business handbook, Beyond Entrepreneurship for the entrepreneurs, visionaries, and innovators of today. This new edition combines the timeless business advice and strategy of the original text, supplemented with cutting-edge insights and case studies pertinent to todays business world.
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <img
            alt="not found"
            class=""
            data-testid="img"
            height="304px"
            src="/Assets/2.png"
            width="304px"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
